carsus/io/cmfgen/base.py

```diff
@@ -0,0 +1,409 @@
+            warnings.warn('Empty table')
+
+        self.fname = fname
+        self.base = df
+        self.columns = df.columns.tolist()
+        self.meta = meta
+
+class CMFGENPhotoionizationCrossSectionParser(BaseParser):
+    """
+        Description
+        ----------
+        base : list of pandas.DataFrame 's
+        columns : list of str
+        meta : dict 
+            Metadata parsed from file header.
+            
+        Methods
+        -------
+        load(fname)
+            Parses the input data and stores the results in the `base` attribute
+    """
+    keys = ['!Date',
+            '!Number of energy levels', 
+            '!Number of photoionization routes', 
+            '!Screened nuclear charge',
+            '!Final state in ion',
+            '!Excitation energy of final state',
+            '!Statistical weight of ion',
+            '!Cross-section unit',
+            '!Split J levels',
+            '!Total number of data pairs',
+            ]
+
+    def _table_gen(self, f):
+        """ Generator. Yields a cross section table for an energy level """
+        meta = {}
+        data = []
+
+        for line in f:
+
+            if '!Configuration name' in line:
+                meta['Configuration'] = line.split()[0]
+
+            if '!Type of cross-section' in line:
+                meta['Type of cross-section'] = int(line.split()[0])
+
+            if '!Number of cross-section points' in line:
+                meta['Points'] = int(line.split()[0])
+
+                p = meta['Points']
+                for i in range(p):
+
+                    values = f.readline().split()
+                    if len(values) == 8:  # Verner ground state fits
+
+                        data.append(list(map(int, values[:2])) + list(map(float, values[2:])))
+
+                        if i == p/len(values) -1:
+                            break
+
+                    else:
+                        data.append(map(to_float, values))
+
+                break
+
+        df = pd.DataFrame.from_records(data)
+        df._meta = meta    
+
+        yield df
+
+
+    def load(self, fname):
+    
+        meta = parse_header(fname, self.keys)
+        tables = []
+        with gzip.open(fname, 'rt') if fname.endswith('.gz') else open(fname) as f :
+
+            while True:
+
+                df = next(self._table_gen(f), None)
+
+                if df.empty:
+                    break
+
+                if df.shape[1] == 2:
+                    df.columns = ['Energy', 'Sigma']
+
+                elif df.shape[1] == 1:
+                    df.columns = ['Fit coefficients']
+
+                elif df.shape[1] == 8:  # Verner ground state fits. TODO: add units
+                    df.columns = ['n', 'l', 'E', 'E_0', 'sigma_0', 'y(a)', 'P', 'y(w)']
+
+                else:
+                    warnings.warn('Inconsistent number of columns')
+
+                tables.append(df)
+
+        tables.insert(0, meta)
```

Comment:
This seems unnecessarily complicated. You insert the meta data into the list; three lines down you separate it out again by slicing the list.

Reply:
Yes. I it was ok when whe haven't any `meta` attribute butis not necessary anymore. Thanks